[ARM64] Implement Duplicate and DuplicateSelectedScalar

tannergooding opened the pull request:
This implements the majority of https://github.com/dotnet/runtime/issues/35037. It does not cover the `InsertSelectedScalar` intrinsics which I will follow up with in a separate PR.

tannergooding committed:
Adding support for DuplicateToVector64 and DuplicateToVector128 to Arm.AdvSimd and Arm.AdvSimd.Arm64

tannergooding committed:
Adding support for DuplicateSelectedScalarToVector64 and DuplicateSelectedScalarToVector128 to Arm.AdvSimd and Arm.AdvSimd.Arm64

tannergooding committed:
Adding tests for DuplicateToVector64 and DuplicateToVector128

tannergooding committed:
Regenerating tests for the newly added HWIntrinsics

tannergooding committed:
Adding tests for DuplicateSelectedScalarToVector64 and DuplicateSelectedScalarToVector128

tannergooding commented:
CC. @CarolEidt, @echesakovMSFT, @kunalspathak

tannergooding committed:
Ensure parentheses exist

kunalspathak commented on src/coreclr/src/jit/hwintrinsiclistarm64.h:
```diff
@@ -192,6 +196,9 @@ HARDWARE_INTRINSIC(AdvSimd_Arm64,   CompareLessThanScalar,
 HARDWARE_INTRINSIC(AdvSimd_Arm64,   Divide,                                    -1,           2,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_fdiv,           INS_fdiv},              HW_Category_SimpleSIMD,             HW_Flag_NoFlag)
+HARDWARE_INTRINSIC(AdvSimd_Arm64,   DuplicateSelectedScalarToVector128,        16,           2,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_dup,            INS_dup,            INS_invalid,        INS_dup},               HW_Category_IMM,                    HW_Flag_SupportsContainment|HW_Flag_BaseTypeFromFirstAr ...
+HARDWARE_INTRINSIC(AdvSimd_Arm64,   DuplicateToVector64,                        8,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_mov,            INS_mov,            INS_invalid,        INS_fmov},              HW_Category_SimpleSIMD,             HW_Flag_SupportsContainment|HW_Flag_SpecialCodeGen)
+HARDWARE_INTRINSIC(AdvSimd_Arm64,   DuplicateToVector128,                      16,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_dup,            INS_dup,            INS_invalid,        INS_dup},               HW_Category_SimpleSIMD,             HW_Flag_SupportsContainment|HW_Flag_SpecialCodeGen)
```
>DuplicateToVector128 [](start = 36, length = 20)

I don't see `DuplicateToVector128` listed under `AdvSimd.Arm64` in [https://github.com/dotnet/runtime/issues/35037](https://github.com/dotnet/runtime/issues/35037).

tannergooding replied:
There was a name mismatch, the API review used `Vector128<ulong>  DuplicateToVector64(ulong value);` when it should have been `Vector128<ulong>  DuplicateToVector128(ulong value);

tannergooding replied:
The `DuplicateToVector64` line shouldn't exist however, I will remove.

tannergooding committed:
Removing the AdvSimd.Arm64 hwintrinsiclist entry for DuplicateToVector64

CarolEidt reviewed (commented):
Minor question & suggestion

kunalspathak commented:
>     ("DuplicateTest.template",       new Dictionary<string, string> { ["TestName"] = "DuplicateToVector64_Byte",                                ["Isa"] = "AdvSimd",       ["LoadIsa"] = "AdvSimd", ["Method"] = "DuplicateToVector64",                     ["RetVectorType"] = "Vector64",  ["RetBaseType"] = "Byte",                                     ["Op1BaseType"] = "Byte",                                                                                                                             ["LargestVectorSize"] = "8",  ["NextValueOp1"] = "TestLibrary.Generator.GetByte()",                                                                                                                                        ["ValidateFirstResult"] = "result[0] != data",                                                                                                                         ["ValidateRemainingResults"] = "result[i] != data"}),

I don't see `DuplicateTest.template` file. Did you forget to add it?

---
Refers to: src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/Shared/GenerateTests.csx:463 in 2057f27. [](commit_id = 2057f27490e56d07fea836e37d5c238aa34ca3e8, deletion_comment = False)

kunalspathak commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -556,14 +585,25 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
-                if (intrin.op1->isContainedFltOrDblImmed())
+            case NI_AdvSimd_DuplicateToVector64:
+            case NI_AdvSimd_DuplicateToVector128:
+            case NI_AdvSimd_Arm64_DuplicateToVector64:
```
>NI_AdvSimd_Arm64_DuplicateToVector64 [](start = 17, length = 36)

I assume you will have to delete this as well since you have remove `DuplicateToVector64` from AdvSimd.ARM64.

tannergooding replied:
Yes. Fixed locally.

tannergooding commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -556,14 +585,25 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                        const double dataValue = intrin.op1->AsDblCon()->gtDconVal;
+                        GetEmitter()->emitIns_R_F(INS_fmov, emitSize, targetReg, dataValue, opt);
+                    }
+                    else if (intrin.id == NI_AdvSimd_Arm64_DuplicateToVector64)
```
This case actually doesn't need to exist, it was the mismatched name from API review that I added and then forgot to remove from the JIT code. It is gone now.

tannergooding committed:
Removing the rest of the NI_AdvSimd_arm64_DuplicateToVector64 usages and clarifying a comment

tannergooding commented:
> I don't see DuplicateTest.template file. Did you forget to include it in PR?

It is a meta template defined here: <https://github.com/dotnet/runtime/pull/36144/files#diff-95804fe9ddc21eeb87df5c86267d76f7R89>
We have a few templates which are used to generate other templates dynamically, to avoid duplicating the central logic when we really only have a couple variances in validation logic or parameter passing.

So `_UnaryOpScalarTestTemplate.template` is the meta template for an instruction which returns a `Vector` and takes a single `T` input operand. `DuplicateTest.template` is generated in memory based on that using `SimpleOpTest_ValidationLogic`

kunalspathak commented:
Ah I see.

---
In reply to: [626046994](https://github.com/dotnet/runtime/pull/36144#issuecomment-626046994) [](ancestors = 626046994)

kunalspathak commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -556,28 +585,33 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+                {
+                    GetEmitter()->emitIns_R_R(ins, emitSize, targetReg, op1Reg, opt);
+                }
                 break;
```
>break; [](start = 16, length = 6)

minor: for consistency, can you put the `break` outside the parenthesis of case?

tannergooding replied:
Fixed.

Is this a convention just for the ARM code? I've not seen it in other parts of the codebase...

kunalspathak replied:
I have seen at few places in this file so that's why thought of keeping it consistent at least in this file. :)

tannergooding committed:
Moving break outside the braces for consistency

tannergooding committed:
Removing another Arm64.DuplicateToVector64

kunalspathak commented:
>             var result = AdvSimd.DuplicateToVector64(data);

I believe this doesn't cover the case  `op1->isContainedIntOrIImmed()`, right? I don't see we have existing coverage for similar scenario in `.Create()` APIs as well.

---
Refers to: src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/AdvSimd/DuplicateToVector64.Byte.cs:195 in 9ccec16. [](commit_id = 9ccec16d16ec8b9f33d9b9b27f1b9e54d9da1b7d, deletion_comment = False)

kunalspathak commented on src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/AdvSimd.Arm64/DuplicateToVector64.Int64.cs:
```diff
@@ -0,0 +1,300 @@
+        {
+            TestLibrary.TestFramework.BeginScenario(nameof(RunBasicScenario_UnsafeRead));
+
+            var result = AdvSimd.Arm64.DuplicateToVector64(
```
>AdvSimd.Arm64.DuplicateToVector64 [](start = 25, length = 33)

We don't have `AdvSimd.Arm64.DuplicateToVector64()` so you don't need this. I am not sure how CI is not complaining. Did you make sure that CI picked up these newly added test?

tannergooding replied:
This was:
* Original change had `DuplicateToVector64`
* Realized it was a typo from API review and isn't something we are exposing
* Tests were regenerated, but the `DuplicateToVector64` tests from the last generation weren't deleted

kunalspathak replied:
Got it. Thanks!

tannergooding replied:
Removed

kunalspathak commented:
>             var result = AdvSimd.DuplicateToVector64(data);

Likewise for other APIs that generates different code if operand is constant.

---
In reply to: [626864715](https://github.com/dotnet/runtime/pull/36144#issuecomment-626864715) [](ancestors = 626864715)

---
Refers to: src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/AdvSimd/DuplicateToVector64.Byte.cs:195 in 9ccec16. [](commit_id = 9ccec16d16ec8b9f33d9b9b27f1b9e54d9da1b7d, deletion_comment = False)

kunalspathak commented on src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/AdvSimd.Arm64/DuplicateToVector64.Double.cs:
```diff
@@ -0,0 +1,300 @@
+        {
+            bool succeeded = true;
+
+            if (result[0] != data)
```
>if (result[0] != data) [](start = 12, length = 22)

curious, why don't we have all the validation inside loop that runs from 0~RetElementCount? see that we use this template at other places but probably since we want to compare all the lanes, just do it from 0~RetElementCount.

tannergooding replied:
Because we didn't use to have meta templates and so we had a single template that handled both `SIMD` and `SIMDScalar` tests. It's just never been normalized yet

kunalspathak commented on src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/AdvSimd/DuplicateSelectedScalarToVector128.V128.Int16.4.cs:
```diff
@@ -0,0 +1,506 @@
+
+            var result = AdvSimd.DuplicateSelectedScalarToVector128(
+                Unsafe.Read<Vector128<Int16>>(_dataTable.inArrayPtr),
+                4
```
>4 [](start = 16, length = 1)

sorry, i might have missed it, but I am not seeing a test where the index element is a non-constant.

tannergooding replied:
Its covered by the reflection scenario

kunalspathak replied:
I see. Is it in some other file? This file still takes `4` in `RunReflectionScenario_UnsafeRead` and `RunReflectionScenario_Load`

kunalspathak replied:
ah, the parameters during invocation will be non-constant I assume.

tannergooding replied:
Yes, reflection requires boxing everything and performing an indirect invocation, so it being a constant is completely lost.

kunalspathak commented:
LGTM, other then my comments in test cases.

kunalspathak reviewed (approved):
:shipit:

tannergooding committed:
Removing the AdvSimd.Arm64.DuplicateToVector64 tests

tannergooding committed:
Adding tests covering DuplicateToVector64/DuplicateToVector128 immediate inputs

tannergooding committed:
Regenerating the ARM64 HWIntrinsic tests

tannergooding commented:
Removed the `DuplicateToVector64` tests since they are dead code and added tests covering the "duplicate immediate" scenario. The tests use 31 as it is a constant that works for both integrals and floating-point values.

CarolEidt reviewed (approved):
LGTM

tannergooding commented:
@kunalspathak, are there not existing tests for `Vector64.Create(immediate)`? The new `DuplicateToVector64(immediate)` tests (and same for `Vector128`) are all failing and I would have expected them to pass since they are using the same path.

tannergooding commented:
Ah, some of the tests are because of a missing cast and so it is calling the `int32` overload rather than `int8` (for example).

kunalspathak commented:
> @kunalspathak, are there not existing tests for `Vector64.Create(immediate)`? The new `DuplicateToVector64(immediate)` tests (and same for `Vector128`) are all failing and I would have expected them to pass since they are using the same path.

Yes, I don't think we have any today and I realized it when I was reviewing this PR. See my previous comment : https://github.com/dotnet/runtime/pull/36144#issuecomment-626864715 . I can see what is going on.

kunalspathak commented:
> Ah, some of the tests are because of a missing cast and so it is calling the `int32` overload rather than `int8` (for example).

Ok. Let me know if there is anything I can help. For some reason it takes a while for me to see the attachments of the test failures in VSO, so still didn't see the actual failure.

tannergooding committed:
Fixing the DuplicateToVector immediate tests

tannergooding committed:
Regenerating the DuplicateToVector immediate tests

tannergooding commented:
Looks like it was entirely due to the missing casts and tests are passing locally. Codegen is similar to:
```asm
4F01F7F0          fmov    v16.4s, #31.0000
```
or
```asm
4F0087F0          movi    v16.8h, #0x1f
```

tannergooding commented:
`long` and `ulong` aren't being optimized correctly however. I logged https://github.com/dotnet/runtime/issues/36254
